### ISSUE TITLE
[Thread] Documentation Requests

### ISSUE THREAD

**Issue by whyrusleeping** [filecoin-project/lotus#2956]
If you have a request for documentation on lotus, lotus-miner, or any related ecosystem tooling, please leave a comment here with what you want documented. 

An example of a good request would be:

> I would like documentation on how to operate multiple lotus-workers on different machines on my local network, with different machines for different jobs

or

> I would like some documentation on how to examine exactly why my message failed on chain.

Please read through existing requests, and thumbs up any requests you also want (think of it like voting) instead of posting a duplicate (though if you want to add something to an existing request, please link to it in your comment, and let us know what you want to add)

**Comment by benjaminh83** [comment]
I would like a change in the documentation on how you compile lotus. Currently is says: 
https://docs.lotu.sh/en+install-lotus-ubuntu#build-the-lotus-binaries-from-source-and-install-185356
but this does not compile from source, and it gives very bad sealing performance (+12h for PC1). I would like it to be: 
`env RUSTFLAGS="-C target-cpu=native -g" FFI_BUILD_FROM_SOURCE=1 make clean all`

**Comment by benjaminh83** [comment]
https://docs.lotu.sh/en+lotus-seal-worker needs an overhaul. Everything needs to be lotus-worker. Almost all commands are old. I would also like to include some usefull commands: 
`export LOTUS_WORKER_PATH=/some/location/for/storage`
and on the lotus-miner, you just run: 
`lotus-miner auth api-info --perm admin`
and you get the token/connection sting that you need to export on the lotus-worker

**Comment by benjaminh83** [comment]
For the lotus-miner, is is super relevant, that you have full control over where sealing and storing happens. documentation needs to be in place for using: 
`lotus-miner storage attach --init --store /some/location/for/long-time-storage`
`lotus-miner storage attach --init --seal /some/location/for/fast-sealing-storage`
And then how to configure the initial storage, if you don't want that to be used: 
`~/.lotusminer/sectorstore.json`
And in general updating sectorstore.json

**Comment by kenshyx** [comment]
I would like to know more about all the sector's lifecycle(priority and resources used) and how to troubleshoot when they are faulty.

**Comment by hxg1987** [comment]
Command set used by miners and common error handling

**Comment by jjxx2021** [comment]
I would like to know how to limit deals count recevied from peers. It could out of capability of all workers.

**Comment by christest1234578** [comment]
I want to know how to set up "1 miner + N worker"

**Comment by litianc** [comment]
I would like documentation on how to recover storage power from faulty and the workflow behind it.

**Comment by bird512** [comment]
I would like to know all the means of the methods mean in the MethodsMiner struct.

**Comment by JayJay1024** [comment]
I would like some documentation on how to obtain/calculate the rewards or earnings of a miner, maybe such as `Filecoin.StateReadState`, `Filecoin.StateMinerAvailableBalance`

**Comment by 815943555** [comment]
Need a guide file for beginners, how to build a cluster step by step

**Comment by ericnjin** [comment]
I want to know how to set up  - multi seal workers / server

**Comment by momack2** [comment]
Update lotu.sh with new storage miner ux

**Comment by stuberman** [comment]
I would like to see a document that focuses on the config.toml files, how they are configured in each location and more detail about the attributes, default values and value ranges. There should also be the ability to prestage a toml file to assist in new builds without the new build overwriting the values.

Examples:

I have a config.toml in LOTUS_PATH which sets the announce address/listener address and port values. When I build a new Lotus instance, instead of getting a random port, I know what it will be and my firewall is ready.

I have a config.toml in LOTUS_STORAGE_PATH which sets the announce address/listener address and port values. When I build a new Lotus instance, instead of getting a random port, I know what it will be and my firewall is ready.
I also disable sealing functions for the miner and let the workers perform sealing functions.

I have a config.toml in WORKER_PATH which sets the API address and port values. When I build a new Lotus instance and attach sealing workers the ports and sealing functions are defined. For workers, I would like to know the importance and default value of the "timeout" attribute as well as the range. Does '0' imply no timeout?

Looking across each TOML file, which sections are used by the Lotus Daemon, the Lotus Miner and the Lotus Worker(s) such that there are no conflicts between files? When using TOML which commands (or arguments) are not needed as you attach paths?

A description of how and when these (daemon, miner and workers) processes read and write to each TOML files is really needed. As an example which attributes will be read as needed versus which attributes are only read when a process first starts? If I alter an attribute do I need to restart the process or will it be read when a phase starts (like PC1, etc)?

**Comment by stuberman** [comment]
Client Deals - a complete description of the steps to make a storage deal. 

I look at it this way today:

1. Find a storage miner you want to deal with - what is a reasonable way of doing this, just random browsing miner IDs in FilFox? _t00000_

2. See if you already have a deal with them: `lotus client list-deals |grep t00000`

3. See if the miner is usable and their terms: `lotus client query-ask t00000`

4. Make an offer: `lotus client deal bafy... t00000 0.005 608000`

5. Check to see the results: `lotus client list-deals |grep t00000`

6. Guide to response from above: 
 6a - Was the deal accepted?
 6b - If there is an error, describe the error types and possible ways to fix the offer

For documentation include the following discussions:
How is pricing calculated, for instance, I made this offer

`lotus client query-ask t00000`

> Ask: t00000
> Price per GiB: 0.001 FIL
> Max Piece size: 100 MiB

lotus client deal bafya... t00000 0.005 608000

...xyz  0       t00000    StorageDealTransferring        N                N            ...abc  3.969 MiB  3050.125 FIL      610025    

How did this deal end up costing 3050.125 FIL?
Since client deals are dependent on Data CIDs, Piece CIDs, and various messages, a guide to each type, how to query against them is essential.

**Comment by stuberman** [comment]
Retrieval Deals - where do I start? There is next to no documentation other than very abstract stuff about the high level handshake process, it needs to go deeper.

Where do I find retrieval candidates? (Today I just look at active storage deals I made: `lotus client list-deals |grep Active`
How do I negotiate a price?

Here is what I am familiar with:
`lotus client retrieve --miner <remote miner> bafy... download/file.output`

And `lotus-miner retrieval-deals list`

And then for a single miner, it can take a very long time to get a response, If I ever get one. If I don't specify a miner, it never seems to work.

Like with the storage deals a document that describe the CID definitions and values is critical.

And I am not sure if this only applies to Space Race, but how is failure/success rate determined?

**Comment by stuberman** [comment]
Run command and [one time initial set up] commands would be helpful for people trying to get started:

[Build commands]
Describe or link to instructions

**Lotus Daemon**
1. Start Lotus daemon - `nohup lotus daemon > ~/lotus.log 2>&1 &`
Check logs - `tail -f lotus.log` or `tail -1000 lotus.log`
If you have not preconfigured your Lotus daemon port: `lotus net listen`
2. Sync Lotus to blockchain `lotus sync wait`
[`lotus wallet new bls `]
[Fund wallet if new and check balance]

**Lotus Miner**
On first start only [`lotus-miner init --no-local-storage --sector-size=32GiB --owner=t3...`]
1. Start Miner `nohup lotus-miner run > ~/miner.log 2>&1 &`
Check logs - `tail -f miner.log` or `tail -1000 miner.log`
*** Wait 5 minutes for miner to start ***
If you have not preconfigured your Lotus miner port: `lotus-miner net listen`
[Put public address on blockchain`lotus-miner actor set-addrs /ip4/<public IP>/tcp/<public port>`]
[Set deal prices `lotus-miner storage-deals set-ask --price n --verified-price n --max-piece-size 32GiB`]
[Specify storage path `lotus-miner storage attach --init --store /mypath/storage`]

**Lotus Worker(s)**
1. Start worker(s) `lotus-worker --worker-repo /Lotus/worker run --listen 127.0.0.1:3456 > ~/worker.log 2>&1 &`
(Provide details to add additional local or remote sealing workers...)

**On going operational guide**
How do you stop or pause a process?
Are there dependencies when trying to stop a process?
What, if anything, needs to be done before you reboot a system?
How do you temporarily stop accepting deals?
What is the impact if you are offline for a little while?
Process to add new hardware components (drives, GPUs, remote workers, etc)

**Comment by Shekelme** [comment]
How can we configure (limit) a quantity of sealing tasks (or a quantity of separate phazes) that can be done simultaneously? My Filecoin server has only 128 GB RAM, and it looks that the best way is to do only one PreCommit1 at a time. But I see often that there are two sectors in P1 state.
Also, if I have several GPUs in the server, can I perform several P2 and C2 tasks?

**Comment by stuberman** [comment]
Documentation on setting up running multiple seal workers concurrently on a single machine - see [Earlier post](https://github.com/filecoin-project/lotus/issues/2956#issuecomment-672876106)
Specifically proper place(s) to put paths (env, json, toml files); proper commands to start multiple worker processes; proper configuration for tasks; and general description of how tasks are ordered and how to allow concurrent processes to run. (I can run two PC1s at the same time with 128 GB ram on AMD 3900xt, but I would like to be able to service other requests at the same time: e.g., run PC1 while running 'UNS' or running PC1 while running C2... using parameters like BELLMAN_CPU_UTILIZATION=0.5 are helpful but there is no unified document to pull the operation view together.

**Comment by stuberman** [comment]
As a lotus miner trying to learn and use the system, an overview of the PAYCH (payment channel) purpose and function would be very helpful. In addition specific commands and use cases would go a long way to helping us understand when or if we need to use this function.

**Comment by whyrusleeping** [comment]
https://filecoinproject.slack.com/archives/CG6P10H9T/p1597418379057900?thread_ts=1597400365.050200&cid=CG6P10H9T

**Comment by Meatball13** [comment]
Would love to see a really good description and resource that lists all the Environmental Variables and flags and what they do/are used for.

**Comment by hsanjuan** [comment]
Just a quick heads up that we're taking note and will create specific tutorials addressing many of the questions here. Some of them will be directly addressed in the next doc iteration (in their respective sections).

**Comment by jennijuju** [comment]
closing this one since most of the request are addressed, thanks to @hsanjuan! Please create a new issue if further doc requests exists.

**Comment by oricot** [comment]
> How can we configure (limit) a quantity of sealing tasks (or a quantity of separate phazes) that can be done simultaneously? My Filecoin server has only 128 GB RAM, and it looks that the best way is to do only one PreCommit1 at a time. But I see often that there are two sectors in P1 state.
lotus-worker run --parallel-fetch-limit=1
(i didn't try it)
